packages/app/src/hooks/useNodeBodyHeight.tsx

```diff
@@ -0,0 +1,59 @@
+import { type NodeId } from '@ironclad/rivet-core';
+import { useCallback, useEffect, useMemo, useRef } from 'react';
+import { useRecoilValue } from 'recoil';
+import { nodesState } from '../state/graph';
+
+export interface NodeHeightCache {
+  get: (nodeId: NodeId) => number | undefined;
+
+  has: (nodeId: NodeId) => boolean;
+
+  set: (nodeId: NodeId, height: number | undefined) => void;
+}
+
+/**
+ * A cache of node heights. This is used when a node is unmounted and moved to the dragging list, since the node's
+ * body needs to be re-rendered in order to get its height. This cache allows us to avoid flickering when the node
+ * is first rendered in the dragging list.
+ */
+export const useNodeHeightCache = (): NodeHeightCache => {
+  const nodes = useRecoilValue(nodesState);
+  const ref = useRef<Record<string, number | undefined>>({});
+
+  const set = useCallback((nodeId: NodeId, height: number | undefined) => {
```

Comment:
In this use case, it's not appropriate to use the `useStableCallback`. As the callbacks defined here are stable and these are not event listener callbacks.

Also, the implementation of `useStableCallback` is (sorry to say) buggy. As the dependencies of the callback weren't added. As a result, that custom hook is probably introducing some bugs elsewhere in the app. It was on my radar to take a closer look at it.

While on that subject, here is a ref to the React documentation explaining their `useEventCallback`:

```
function useEventCallback(fn, dependencies) {
  const ref = useRef(() => {
    throw new Error('Cannot call an event handler while rendering.');
  });

  useEffect(() => {
    ref.current = fn;
  }, [fn, ...dependencies]);

  return useCallback(() => {
    const fn = ref.current;
    return fn();
  }, [ref]);
}
```

Above, we can see the `dependencies` is used to trigger `useEffect` when the callback function mutates. So why are they doing this? Well, it mutates the "ref" instead of triggering prop drilling by any components using the callback. Thus, components using the callback remain stable while the `useRef` gets updated with the callback.

This is really important in their example, that this pattern can only work with DOM event listeners. As updating of the `useRef` is lagging behind render updates.

We've used this custom hook in many places, in particular, with callbacks that had dependencies which weren't defined and in cases where the callbacks aren't event listeners. So I feel we need to go through the code and clean up the usages, and as a side effect will probably fix some bugs which we weren't tracking.

FYI, the bug I'm referring to is here in our code `useUnrenderedValue.ts`:

```
  useLayoutEffect(() => {
    unrenderedValueRef.current.value = value;
  });
```

This becomes a 1 time update of the ref with the original callback and that's not correct. It's a very easy fix (not a big deal) and we can address this with another PR later.

Reply:
Interesting! @ZhangYiJiang I think you wrote it what's your take?

Reply:
> This is really important in their example, that this pattern can only work with DOM event listeners. 

Yeah this is what useStableCallback is specifically for (it could also be used in useEffect or any sort of side effect). There is in fact a proposal to roll something similar into React core called useEvent although I think that did not make it. Our version could use a better name but this is in fact a feature not a bug

Reply:
was I using it in an effect? my bad